nowcasting_dataset/data_sources/sun/raw_data_load_save.py

```diff
@@ -0,0 +1,181 @@
+) -> (pd.DataFrame, pd.DataFrame):
+    """
+
+    Get Azimuth and elevation positions of the sun
+
+    For a list of datestamps and a list of coordinates, get the azimuth and elevation degrees.
+    Note that the degrees are rounded to 2 decimal places, as we at most need that.
+
+    Args:
+        datestamps: list of datestamps that are needed
+        x_centers: list of x coordinates - ref. OSGB
+        y_centers: list of y coordinates - ref. OSGB
+
+    Returns: Tuple of dataframes for azimuth and elevation.
+        The index is timestamps, and the columns are the x and y coordinates
+
+    """
+
+    logger.debug(
+        f"Will be calculating for {len(datestamps)} datestamps and {len(x_centers)} locations"
+    )
+
+    # create array of index datetime, columns of system_id for both azimuth and elevation
+    azimuth = []
+    elevation = []
+
+    t = time.time()
+    names = []
+    # loop over locations and find azimuth and elevation angles,
+    with futures.ThreadPoolExecutor() as executor:
+
+        logger.debug("Setting up jobs")
+
+        # Submit tasks to the executor.
+        future_azimuth_and_elevation_per_location = []
+        for i in tqdm(range(len(x_centers))):
+
+            name = x_y_to_name(x_centers[i], y_centers[i])
+            if name not in names:
+
+                lat, lon = geospatial.osgb_to_lat_lon(x=x_centers[i], y=y_centers[i])
+
+                future_azimuth_and_elevation = executor.submit(
+                    geospatial.calculate_azimuth_and_elevation_angle,
+                    latitude=lat,
+                    longitude=lon,
+                    datestamps=datestamps,
+                )
+                future_azimuth_and_elevation_per_location.append(
+                    [future_azimuth_and_elevation, name]
+                )
+                names.append(name)
+
+        logger.debug(f"Getting results")
+
+        # Collect results from each thread.
+        for i in tqdm(range(len(future_azimuth_and_elevation_per_location))):
+            future_azimuth_and_elevation, name = future_azimuth_and_elevation_per_location[i]
+            azimuth_and_elevation = future_azimuth_and_elevation.result()
+
+            azimuth_per_location = azimuth_and_elevation.loc[:, "azimuth"].rename(name)
+            elevation_per_location = azimuth_and_elevation.loc[:, "elevation"].rename(name)
+
+            azimuth.append(azimuth_per_location)
+            elevation.append(elevation_per_location)
+
+    azimuth = pd.concat(azimuth, axis=1)
+    elevation = pd.concat(elevation, axis=1)
+
+    # remove timezone
+    elevation.index = elevation.index.tz_localize(None)
+    azimuth.index = azimuth.index.tz_localize(None)
+
+    logger.debug(f"Calculated Azimuth and Elevation angles in {time.time() - t} seconds")
+
+    return azimuth.round(2), elevation.round(2)
+
+
+def save_to_zarr(azimuth: pd.DataFrame, elevation: pd.DataFrame, filename: Union[str, Path]):
```

Comment:
If we almost always read the entire dataset from disk at once, then it may be faster to read a single large file rather than reading lots of small files (an, under the hood, Zarr is lots of small files).  Maybe NetCDF or HDF?

But this isn't super-important!

Reply:
yea good point - perhaps I''l leave that for another time